p2p/channel.go

```diff
@@ -80,86 +81,218 @@ type stream struct {
+	proxyPort, err := port.NewPool().Acquire()
 	if err != nil {
-		return nil, fmt.Errorf("could not create UDP conn: %w", err)
+		return nil, err
+	}
+	proxyConn, err := net.ListenUDP("udp4", &net.UDPAddr{IP: net.ParseIP("127.0.0.1"), Port: proxyPort.Num()})
+	if err != nil {
+		return nil, err
 	}
 
-	peerAddr := punchedConn.RemoteAddr().(*net.UDPAddr)
-	udpSession, err := kcp.NewConn3(1, peerAddr, blockCrypt, 10, 3, udpConn)
+	// Setup KCP session. It will write to proxy conn only.
+	udpSession, sessAddr, err := listenUDPSession(proxyConn.LocalAddr(), privateKey, peerPubKey)
 	if err != nil {
-		return nil, fmt.Errorf("could not create UDP session: %w", err)
+		return nil, fmt.Errorf("could not create KCP UDP session: %w", err)
 	}
 
 	tr := transport{
-		Reader:  &textproto.Reader{R: bufio.NewReader(udpSession)},
-		Writer:  &textproto.Writer{W: bufio.NewWriter(udpSession)},
-		session: udpSession,
-		conn:    udpConn,
+		textReader: textproto.NewReader(bufio.NewReader(udpSession)),
+		textWriter: textproto.NewWriter(bufio.NewWriter(udpSession)),
+		session:    udpSession,
+		remoteConn: remoteConn,
+		proxyConn:  proxyConn,
+	}
+
+	peer := peer{
+		publicKey:  peerPubKey,
+		remoteAddr: peerAddr,
 	}
 
 	c := channel{
-		tr:            &tr,
-		topicHandlers: make(map[string]HandlerFunc),
-		streams:       make(map[uint64]*stream),
-		privateKey:    privateKey,
-		peerPubKey:    peerPubKey,
-		peerAddr:      peerAddr,
-		localAddr:     udpConn.LocalAddr().(*net.UDPAddr),
-		blockCrypt:    blockCrypt,
-		serviceConn:   nil,
-		stop:          make(chan struct{}, 1),
-		sendQueue:     make(chan *transportMsg, 100),
+		tr:               &tr,
+		topicHandlers:    make(map[string]HandlerFunc),
+		streams:          make(map[uint64]*stream),
+		privateKey:       privateKey,
+		peer:             &peer,
+		localSessionAddr: sessAddr,
+		serviceConn:      nil,
+		stop:             make(chan struct{}, 1),
+		sendQueue:        make(chan *transportMsg, 100),
 	}
 
-	go c.readLoop()
-	go c.sendLoop()
+	go c.remoteReadLoop()
+	go c.remoteSendLoop()
+	go c.localReadLoop()
+	go c.localSendLoop()
 
 	return &c, nil
 }
 
-// readLoop reads incoming requests or replies to initiated requests.
-func (c *channel) readLoop() {
+// remoteReadLoop reads from remote conn and writes to local KCP UDP conn.
+// If remote peer addr changes it will be updated and next send will use new addr.
+func (c *channel) remoteReadLoop() {
+	// TODO: Check on buf size. Current size 1500 is max MTU as it is also used in KCP.
+	buf := make([]byte, 1500)
+	latestPeerAddr := c.peer.addr()
+	for {
+		select {
+		case <-c.stop:
+			return
+		default:
+		}
+
+		n, addr, err := c.tr.remoteConn.ReadFrom(buf)
+		if err != nil {
+			log.Warn().Err(err).Msgf("Read from remote conn failed")
+			// TODO: Check if exit is ok here.
+			return
+		}
+
+		// Check if peer address is not changed and update it if changed.
+		if addr, ok := addr.(*net.UDPAddr); ok {
+			if !addrEqual(latestPeerAddr, addr) {
```

Comment:
This could be very insecure. I should only update port but not the whole IP. It need to be checked if addr.IP is the same as initial peer addr.IP.